mlir/lib/Bindings/Python/IRCore.cpp

```diff
@@ -3425,19 +3425,35 @@ void mlir::python::populateIRCore(py::module &m) {
           kValueDunderStrDocstring)
       .def(
           "get_name",
-          [](PyValue &self, bool useLocalScope) {
+          [](PyValue &self, std::optional<bool> useLocalScope,
+             std::optional<std::reference_wrapper<PyAsmState>> state) {
             PyPrintAccumulator printAccum;
-            MlirOpPrintingFlags flags = mlirOpPrintingFlagsCreate();
-            if (useLocalScope)
-              mlirOpPrintingFlagsUseLocalScope(flags);
-            MlirAsmState state = mlirAsmStateCreateForValue(self.get(), flags);
-            mlirValuePrintAsOperand(self.get(), state, printAccum.getCallback(),
+            MlirOpPrintingFlags flags;
+            MlirAsmState valueState;
+            // Use state if provided, else create a new state.
+            if (state) {
+              valueState = state.value().get().get();
+              // Don't allow setting using local scope and state at same time.
+              if (useLocalScope)
```

Comment:
won't this be truthy even if user passes in a `False` (since the `std::optional` will actually contain a `bool` value instead of `nullopt`)? Maybe `std::optional` delegates to the truthiness of the contained value when there is one...

Reply:
yea...

```cpp
#include <iostream>
#include <optional>

int main() {
    std::optional<bool> useLocalScope;

    if (useLocalScope)
        std::cout << "empty" << std::endl;

    useLocalScope = false;
    if (useLocalScope)
        std::cout << "false-positive" << std::endl;

    useLocalScope = true;
    if (useLocalScope)
        std::cout << "positive-positive" << std::endl;

    return 0;
}
```
prints 

```
false-positive
positive-positive
```

Reply:
That is the intention though: this should check not the value of the flag, but whether it was set. Let me add has_value() to make that clearer.

Reply:
Aww jeez my bad this just leads to an exception - that's what I get for "peep hole review"